caustic/namespace_dict.py

```diff
@@ -0,0 +1,178 @@
+        flattened_dict = NamespaceDict()
+        def _flatten_dict(dictionary, parent_key=""):
+            for key, value in dictionary.items():
+                if isinstance(value, dict):
+                    _flatten_dict(value, key)
+                else:
+                    flattened_dict[key] = value
+        _flatten_dict(self)
+        return flattened_dict
+
+
+class _NestedNamespaceProxy(_NestedNamespaceDict):
+    """
+    Proxy for NestedNamespaceDict in order to allow recursion in
+    the class attributes
+    """
+    def __init__(self, parent, key_path):
+        # Add new private keys to give us a ladder back to root node
+        self._parent = parent
+        self._key_path = key_path
+        super().__init__(parent[key_path])
+
+    def __setattr__(self, key, value):
+        if key.startswith('_'):
+            # We are in a child node, we need to recurse up
+            super().__setattr__(key, value)
+        else:
+            # We are at the root node, call the __setitem__ to set record value
+            self._parent.__setitem__(f"{self._key_path}.{key}", value)
+
+    # Hide the private keys from common usage
+    def keys(self):
+        return [key for key in super().keys() if not key.startswith("_")]
+            
+    def items(self):
+        for key, value in super().items():
+            if not key.startswith('_'):
+                yield (key, value)
+
+    def values(self):
+        return [v for k, v in super().items() if not k.startswith("_")]
+    
+    def __len__(self):
+        # make sure hidden keys don't count in the length of the object
+        return len(self.keys())
+
+
+class NestedNamespaceDict(_NestedNamespaceDict):
+    """
+    Example usage
+    ```python 
+        nested_namespace = NestedNamespaceDict()
+        nested_namespace.foo = 'Hello'
+        nested_namespace.bar = {'baz': 'World'}
+        nested_namespace.bar.qux = 42
+        # works also in the follwoing way
+         nested_namespace["bar.qux"] = 42
+
+        print(nested_namespace)
+        # Output:
+        # {'foo': 'Hello', 'bar': {'baz': 'World', 'qux': 42 }}
+        
+        #==============================
+        # Flattened key access
+        #==============================
+        print(nested_dict['foo'])        # Output: Hello
+        print(nested_dict['bar.baz'])    # Output: World
+        print(nested_dict['bar.qux'])    # Output: 42
+        
+        #==============================
+        # Nested namespace access
+        #==============================
+        print(nested_dict.foo.bar.qux)       # Output: 42
+        
+        #==============================
+        # Flatten and collapse method
+        #==============================
+        print(nested_dict.flatten())
+        # Output:
+        # {'foo': 'Hello', 'foo.bar.baz': 'World', 'foo.bar.qux': 42}
+        
+        print(nested_dict.collapse()
+        # Output:
+        # {'foo': 'Hello', 'baz': 'World', 'qux': 42}
```

Comment:
The purpose of collapse is precisely to create collision, but is not used yet in the code. I was thinking it would be useful if we have DAG structure that enforces unique name for each of its module. A collision, in such a case, would allow us to resolve a node with multiple edges as a unique object.